src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderUtilities.cs

```diff
@@ -76,36 +81,45 @@ internal static bool ContainsNonAscii(string input)
         // encoding'lang'PercentEncodedSpecials
         internal static string Encode5987(string input)
         {
-            // Encode a string using RFC 5987 encoding.
-            // encoding'lang'PercentEncodedSpecials
             var builder = new ValueStringBuilder(stackalloc char[256]);
             byte[] utf8bytes = ArrayPool<byte>.Shared.Rent(Encoding.UTF8.GetMaxByteCount(input.Length));
             int utf8length = Encoding.UTF8.GetBytes(input, 0, input.Length, utf8bytes, 0);
 
             builder.Append("utf-8\'\'");
-            for (int i = 0; i < utf8length; i++)
+
+            ReadOnlySpan<byte> utf8 = utf8bytes.AsSpan(0, utf8length);
+            do
             {
-                byte utf8byte = utf8bytes[i];
+                int length = utf8.IndexOfAnyExcept(s_rfc5987AttrBytes);
+                if (length < 0)
+                {
+                    length = utf8.Length;
+                }
+
+                Encoding.ASCII.GetChars(utf8.Slice(0, length), builder.AppendSpan(length));
```

Comment:
Should this really be `Encoding.ASCII`?

Given that the failure is on alpine I wonder if it could be some locale issue?

Reply:
It shouldn't be affected by locale. The input here will always be these bytes (all in the `[0, 127]` range).
https://github.com/dotnet/runtime/blob/2ae7f5d42c817614839f3bf2e889981fdaa03631/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderUtilities.cs#L24-L27

Reply:
Makes sense, thanks. I also just noticed that there are non-alpine failures too.

Reply:
https://runfo.azurewebsites.net/search/tests/?q=started%3A%7E7+definition%3Aruntime+name%3A%22FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly%22

Alpine and Ubuntu (both arm64)